src/main/java/org/variantsync/diffdetective/feature/BooleanAbstraction.java

```diff
@@ -132,61 +187,62 @@ public String applyTo(String value) {
-
-    private static String abstractAll(String formula, final List<Replacement> replacements) {
-        for (var replacement : replacements) {
+    /**
+     * Apply all possible abstraction replacements for substrings of the given formula.
+     * @param formula the formula to abstract
+     * @return a fully abstracted formula
+     */
+    public static String abstractAll(String formula) {
+        for (var replacement : BooleanAbstraction.REPLACEMENTS) {
             formula = replacement.applyTo(formula);
         }
         return formula;
     }
 
     /**
-     * Abstracts all arithmetics in the given formula.
-     * For example, a formula "3 >= 1 + 2" would be abstracted to a single variable "3__GEQ__1__ADD__2".
-     * The given formula should be a string of a CPP conforming condition.
-     * @param formula The formula whose arithmetics should be abstracted.
-     * @return A copy of the formula with abstracted arithmetics.
-     */
-    public static String arithmetics(final String formula) {
-        return abstractAll(formula, ARITHMETICS);
-    }
-
-    /**
-     * Abstracts parentheses, including the commas of macro calls, in the given formula.
-     *
-     * For example, a call "FOO(3, 4, lol)" would be abstracted to a single variable "FOO__3__4__lol".
-     * The given formula should be a string of a CPP conforming condition.
-     * @param formula The formula whose function calls should be abstracted.
-     * @return A copy of the formula with abstracted function calls.
+     * Search for the first replacement that matches the entire formula and apply it. If no
+     * replacement for the entire formula is found, all possible replacements are applied to
+     * substrings of the formula.
+     * @param formula the formula to abstract
+     * @return a fully abstracted formula
      */
-    public static String parentheses(String formula) {
-        ////// abstract function calls
-        /// replace commata in macro calls
-        formula = COMMA.matcher(formula).replaceAll(COMMA_REPLACEMENT);
-
-        /// inline macro calls as long as there are some
-        /// Example
-        ///    bar(2, foo(A__MUL__(B__PLUS__C))
-        /// -> bar(2__foo(A__MUL__(B__PLUS__C))) // because of the comma replacement above
-        /// -> bar(2__foo(A__MUL____LB__B__PLUS__C__RB__))
-        /// -> bar(2__foo__LB__A__MUL____LB__B__PLUS__C__RB____RB__)
-        /// -> bar__LB__2__foo__LB__A__MUL____LB__B__PLUS__C__RB____RB____RB__
-        String old;
-        do {
-            old = formula;
-            formula = CALL.matcher(formula).replaceAll(CALL_REPLACEMENT);
-//            formula = formula.replaceAll("(\\w+)\\((\\w*)\\)", "$1__$2");
-        } while (!old.equals(formula));
-
-        return formula;
+    public static String abstractFirstOrAll(String formula) {
+        for (Replacement replacement : REPLACEMENTS) {
+            if (replacement.pattern.pattern().equals(formula)) {
+                return replacement.replacement;
```

Comment:
Is this correct? Why isn't `Replacement::applyTo` used here?

Reply:
Good catch. It currently works as intended, because the replacement always corresponds to a specific token. However, this implementation might cause bugs in the future. I changed the method to first check whether the pattern matches the entire formula, and apply the replacement if it is the case.

The intention of the method is to achieve a shorter runtime in case of single tokens that need abstraction. 
I added additional documentation, to explain this intention. On a side node, it could be that the entire method is a case of premature optimization, which is not required.

Reply:
Ok, could you document an example too where such an optimized replacement is valid? I.e., how to know when to use this method and when not? I have no clue right now.

Reply:
Done